[WebCore]:Assign the scalar by default to avoid potential function calls

LumioseSil opened the pull request:
#### 461e6c05f16d62178184cb0fd4070aa5148c6465
<pre>
[WebCore]:Assign the scalar by default to avoid potential function calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=257421">https://bugs.webkit.org/show_bug.cgi?id=257421</a>

Reviewed by Tim Nguyen.

We can avoid calls to functions like targetX and targetY if we assign
the scalar first and then do checking of attributes.

* Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
(WebCore::SVGFEConvolveMatrixElement::attributeChanged): Fixed space typo.
(WebCore::SVGFEConvolveMatrixElement::createFilterEffect const): Refactored.

Canonical link: <a href="https://commits.webkit.org/264641@main">https://commits.webkit.org/264641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/923e3f65a60ceb65b4962828eff197d44e34304e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8486 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8703 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9858 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8260 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8202 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10474 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8397 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11136 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9407 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7433 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9980 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6707 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15053 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7628 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10983 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6582 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7390 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1979 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11597 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7840 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
```diff
@@ -216,54 +216,56 @@ void SVGFEConvolveMatrixElement::svgAttributeChanged(const QualifiedName& attrNa
         targetYValue = static_cast<int>(floorf(orderYValue / 2));
+    else
+        targetYValue = targetY();
 
```
The logic here isn't equivalent, you inverted the if / else here.

Side note: I think this would look better written as a ternary:

```
// The spec says the default value is: targetX = floor ( orderX / 2 ))
int targetXValue = hasAttribute(SVGNames::targetXAttr) ? targetX() : static_cast<int>(floorf(orderXValue / 2));

// The spec says the default value is: targetY = floor ( orderY / 2 ))
int targetYValue = hasAttribute(SVGNames::targetYAttr) ? targetY() : static_cast<int>(floorf(orderYValue / 2));

AZero13 replied:
Fixed!

nt1m commented on Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
```diff
@@ -216,54 +216,47 @@ void SVGFEConvolveMatrixElement::svgAttributeChanged(const QualifiedName& attrNa
+    int orderXValue = 3;
+    int orderYValue = 3;
+    if (hasAttribute(SVGNames::orderAttr)) {
+        orderXValue = orderX();
+        orderYValue = orderY();
+        // Spec says order must be > 0. Bail if it is not.
+        if (orderXValue <= 0 || orderYValue <= 0)
+            return nullptr;
     }
-    // Spec says order must be > 0. Bail if it is not.
-    if (orderXValue < 1 || orderYValue < 1)
-        return nullptr;
```
After reviewing this more carefully, this logic isn't equivalent either. We were early returning unconditionally before, now we're only early returning when the attribute is set.

Can you move out the early return of the `hasAttribute(SVGNames::orderAttr))` check?

The code should look like this:

```
     int orderXValue = 3;
     int orderYValue = 3;
     if (hasAttribute(SVGNames::orderAttr)) {
         orderXValue = orderX();
         orderYValue = orderY();
     }
     // Spec says order must be > 0. Bail if it is not.
     if (orderXValue <= 0 || orderYValue <= 0)
         return nullptr;
```

AZero13 replied:
Actually it is not, because it is impossible to return when the condition is not set.

3 <= 0 is always false.

AZero13 replied:
If the attribute is not set, we will NEVER return unless some outside force messes with the memory or CPU or something, which even in this case we don't want to do.

nt1m replied:
Ah right, I missed that.

nt1m replied:
Fwiw, in terms of performance, I don't think this really matters, given the compiler should be smart enough to optimize this.

nt1m commented on Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
```diff
@@ -216,54 +216,47 @@ void SVGFEConvolveMatrixElement::svgAttributeChanged(const QualifiedName& attrNa
-        kernelUnitLengthYValue = 1;
+    // The spec says the default kernelUnitLength is 1.0, and a specified length cannot be 0.
+    int kernelUnitLengthXValue = 1;
+    int kernelUnitLengthYValue = 1;
+    if (hasAttribute(SVGNames::kernelUnitLengthAttr)) {
+        kernelUnitLengthXValue = kernelUnitLengthX();
+        kernelUnitLengthYValue = kernelUnitLengthY();
+        if (kernelUnitLengthXValue <= 0 || kernelUnitLengthYValue <= 0)
+            return nullptr;
     }
-    if (kernelUnitLengthXValue <= 0 || kernelUnitLengthYValue <= 0)
-        return nullptr;
```
Same here.

AZero13 replied:
Ditto.

LumioseSil committed:
[WebCore]:Assign the scalar by default to avoid potential function calls
https://bugs.webkit.org/show_bug.cgi?id=257421

Reviewed by Tim Nguyen.

We can avoid calls to functions like targetX and targetY if we assign
the scalar first and then do checking of attributes.

* Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
(WebCore::SVGFEConvolveMatrixElement::attributeChanged): Fixed space typo.
(WebCore::SVGFEConvolveMatrixElement::createFilterEffect const): Refactored.

Canonical link: https://commits.webkit.org/264641@main
